transport/tchannel/handler.go

```diff
@@ -188,21 +189,28 @@ func (h handler) callHandler(ctx context.Context, call inboundCall, responseWrit
 type responseWriter struct {
 	failedWith         error
 	format             tchannel.Format
-	headers            transport.Headers
+	headers            map[string]string
 	buffer             *bufferpool.Buffer
 	response           inboundCallResponse
 	isApplicationError bool
+	exactCaseHeader    bool
 }
 
-func newResponseWriter(response inboundCallResponse, format tchannel.Format) *responseWriter {
+func newResponseWriter(response inboundCallResponse, format tchannel.Format, exactCaseHeader bool) *responseWriter {
 	return &responseWriter{
-		response: response,
-		format:   format,
+		response:        response,
+		format:          format,
+		headers:         map[string]string{},
+		exactCaseHeader: exactCaseHeader,
 	}
 }
 
 func (rw *responseWriter) AddHeaders(h transport.Headers) {
-	for k, v := range h.Items() {
+	headers := h.Items()
+	if rw.exactCaseHeader {
+		headers = h.OriginalItems()
+	}
+	for k, v := range headers {
 		// TODO: is this considered a breaking change?
 		if isReservedHeaderKey(k) {
```

Comment:
Note: @kriskowal @abhinav  with exact items, this likely won't be hit if the user uses a different case, any concerns?